### PR TITLE
feat: add rater for mono vertex

### PR DESCRIPTION
**Pull request by kohlisid**

### CODE REVIEW

File: pkg/daemon/server/service/rater/timestamped_counts.go [hunk 1]
```diff
@@ -33,7 +33,7 @@ type TimestampedCounts struct {
 
 func NewTimestampedCounts(t int64) *TimestampedCounts {
 	return &TimestampedCounts{
-		timestamp:         t,
```
Comment (KeranYang) [inline pkg/daemon/server/service/rater/timestamped_counts.go:36]: Do we need to expose the timestamp to the outside of the rater pkg?

File: pkg/metrics/metrics.go [hunk 2]
```diff
@@ -22,16 +22,17 @@ import (
...
+	LabelPlatform               = "platform"
+	LabelNamespace              = "ns"
+	LabelISBService             = "isbsvc"
+	LabelPipeline               = "pipeline"
+	LabelVertex                 = "vertex"
+	LabelMonoVertex             = "mvtx"
+	LabelVertexReplicaIndex     = "replica"
+	LabelVertexType             = "vertex_type"
+	LabelPartitionName          = "partition_name"
+	LabelMonoVertexReplicaIndex = "mvtx-replica"
```
Comment (KeranYang) [inline pkg/metrics/metrics.go:35]: nit:
```suggestion
	LabelVertexReplicaIndex     = "replica"
	LabelVertexType             = "vertex_type"
	LabelPartitionName          = "partition_name"
	LabelMonoVertex             = "mvtx"
	LabelMonoVertexReplicaIndex = "mvtx_replica"
```

File: pkg/mvtxdaemon/server/daemon_server.go [hunk 3]
```diff
@@ -59,6 +59,8 @@ func (ds *daemonServer) Run(ctx context.Context) error {
 	var (
 		err error
 	)
+	// rater is used to calculate the processing rate for each of the vertices
```
Comment (KeranYang) [inline pkg/mvtxdaemon/server/daemon_server.go:62]:
```suggestion
	// rater is used to calculate the processing rate of the mono vertex
```

File: pkg/mvtxdaemon/server/service/mvtx_service.go [hunk 4]
```diff
@@ -27,26 +27,29 @@ import (
...
+	raterPkg "github.com/numaproj/numaflow/pkg/mvtxdaemon/server/service/rater"
 	"github.com/numaproj/numaflow/pkg/shared/logging"
-	"github.com/prometheus/common/expfmt"
 )
 
 type MoveVertexService struct {
 	mvtxdaemon.UnimplementedMonoVertexDaemonServiceServer
 	monoVtx    *v1alpha1.MonoVertex
 	httpClient *http.Client
 	// TODO: add rater
```
Comment (KeranYang) [inline pkg/mvtxdaemon/server/service/mvtx_service.go:43]:
```suggestion
```

File: pkg/mvtxdaemon/server/service/mvtx_service.go [hunk 5]
```diff
@@ -75,11 +79,12 @@ func (mvs *MoveVertexService) getPending(ctx context.Context) map[string]*wrappe
 	log := logging.FromContext(ctx)
 	headlessServiceName := mvs.monoVtx.GetHeadlessServiceName()
 	pendingMap := make(map[string]*wrapperspb.Int64Value)
+	monovtxPendingMetric := "monovtx_pending"
```
Comment (KeranYang) [inline pkg/mvtxdaemon/server/service/mvtx_service.go:82]: Should this be a constant and grouped with other metric names?

File: pkg/mvtxdaemon/server/service/rater/podtracker.go
Comment (KeranYang) [inline pkg/mvtxdaemon/server/service/rater/podtracker.go:1]: naming: `pod_tracker.go`

File: pkg/mvtxdaemon/server/service/rater/rater.go [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+		log.Debugf("Pod %s does not exist, updating it with nil...", pInfo.podName)
+		podReadCount = nil
+	}
+	now := time.Now().Add(CountWindow).Truncate(CountWindow).Unix()
+	UpdateCount(r.timestampedPodCounts, now, podReadCount)
+	return nil
+}
+
+// getPodReadCounts returns the total number of messages read by the pod
+// since a pod can read from multiple partitions, we will return a map of partition to read count.
```
Comment (KeranYang) [inline pkg/mvtxdaemon/server/service/rater/rater.go:136]: please update the comment.

File: pkg/mvtxdaemon/server/service/rater/rater_test.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rater
+
+//type raterMockHttpClient struct {
```
Comment (KeranYang) [inline pkg/mvtxdaemon/server/service/rater/rater_test.go:19]: TODO

File: pkg/mvtxdaemon/server/service/rater/timestamped_counts.go [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+import (
+	"fmt"
+	"sync"
+)
+
+// TimestampedCounts track the total count of processed messages for a list of pods at a given timestamp
+type TimestampedCounts struct {
+	// Timestamp in seconds is the time when the count is recorded
+	Timestamp int64
```
Comment (KeranYang) [inline pkg/mvtxdaemon/server/service/rater/timestamped_counts.go:27]: same comment as above, do we need to expose?

File: rust/monovertex/src/forwarder.rs [hunk 9]
```diff
@@ -40,9 +34,7 @@ impl Forwarder {
                 MONO_VERTEX_NAME.to_string(),
                 config().mono_vertex_name.clone(),
             ),
-            (VERTEX_TYPE_LABEL.to_string(), MONO_VERTEX_TYPE.to_string()),
             (REPLICA_LABEL.to_string(), config().replica.to_string()),
```
Comment (KeranYang) [inline rust/monovertex/src/forwarder.rs:37]: For the replica label, we are reusing `replica` here but we also defined a separate one for the mono vertex above, which is `mvtx_replica`, should it be consistent?

File: rust/monovertex/src/metrics.rs [hunk 10]
```diff
@@ -17,22 +17,30 @@ use tokio::task::JoinHandle;
...
+use prometheus_client::registry::Registry;
+
 use crate::error::Error;
 use crate::sink::SinkClient;
 use crate::source::SourceClient;
 use crate::transformer::TransformerClient;
 
 // Define the labels for the metrics
-pub const MONO_VERTEX_NAME: &str = "vertex";
+pub const MONO_VERTEX_NAME: &str = "mono_vertex_name";
```
Comment (KeranYang) [inline rust/monovertex/src/metrics.rs:34]:
```suggestion
pub const MONO_VERTEX_NAME_LABEL: &str = "mono_vertex_name";
```

also should it be `mvtx`? to be consistent with what we defined in GO? `LabelMonoVertex             = "mvtx"`

File: rust/monovertex/src/metrics.rs [hunk 11]
```diff
@@ -17,22 +17,30 @@ use tokio::task::JoinHandle;
...
 pub const REPLICA_LABEL: &str = "replica";
```
Comment (KeranYang) [inline rust/monovertex/src/metrics.rs:35]: `mvtx_replica`?

**Comment by kohlisid** [comment]
```
{
  "metrics": {
    "monoVertex": "simple-mono-vertex",
    "processingRates": {
      "15m": 48267.2413793104,
      "1m": 48058.3333333333,
      "5m": 48267.2413793104,
      "default": 48212.5
    },
    "pendings": {
      "15m": "254",
      "1m": "200",
      "5m": "270"
    }
  }
}
```

**Comment by kohlisid** [comment]
```
# HELP monovtx_read A Counter to keep track of the total number of messages read from the source.
# TYPE monovtx_read counter
monovtx_read_total{mvtx_name="simple-mono-vertex",mvtx_replica="0"} 17180500
# HELP monovtx_sink_write A Counter to keep track of the total number of messages written to the sink.
# TYPE monovtx_sink_write counter
monovtx_sink_write_total{mvtx_name="simple-mono-vertex",mvtx_replica="0"} 17180000
# HELP monovtx_ack A Counter to keep track of the total number of messages acknowledged by the sink.
# TYPE monovtx_ack counter
monovtx_ack_total{mvtx_name="simple-mono-vertex",mvtx_replica="0"} 17180000
# HELP monovtx_processing_time A Histogram to keep track of the total time taken to forward a chunk, the time is in microseconds.
# TYPE monovtx_processing_time histogram
monovtx_processing_time_sum{mvtx_name="simple-mono-vertex",mvtx_replica="0"} 355918260.0
monovtx_processing_time_count{mvtx_name="simple-mono-vertex",mvtx_replica="0"} 34360
monovtx_processing_time_bucket{le="100.0",mvtx_name="simple-mono-vertex",mvtx_replica="0"} 0
monovtx_processing_time_bucket{le="90000000000.0",mvtx_name="simple-mono-vertex",mvtx_replica="0"} 34360
monovtx_processing_time_bucket{le="81000000000000000000.0",mvtx_name="simple-mono-vertex",mvtx_replica="0"} 34360
monovtx_processing_time_bucket{le="7.2899999999999995e28",mvtx_name="simple-mono-vertex",mvtx_replica="0"} 34360
monovtx_processing_time_bucket{le="6.560999999999999e37",mvtx_name="simple-mono-vertex",mvtx_replica="0"} 34360
monovtx_processing_time_bucket{le="5.9049e46",mvtx_name="simple-mono-vertex",mvtx_replica="0"} 34360
monovtx_processing_time_bucket{le="5.31441e55",mvtx_name="simple-mono-vertex",mvtx_replica="0"} 34360
monovtx_processing_time_bucket{le="4.7829689999999998e64",mvtx_name="simple-mono-vertex",mvtx_replica="0"} 34360
monovtx_processing_time_bucket{le="4.3046721e73",mvtx_name="simple-mono-vertex",mvtx_replica="0"} 34360
monovtx_processing_time_bucket{le="3.87420489e82",mvtx_name="simple-mono-vertex",mvtx_replica="0"} 34360
monovtx_processing_time_bucket{le="+Inf",mvtx_name="simple-mono-vertex",mvtx_replica="0"} 34360
# HELP monovtx_read_bytes A Counter to keep track of the total number of bytes read from the source.
# TYPE monovtx_read_bytes counter
monovtx_read_bytes_total{mvtx_name="simple-mono-vertex",mvtx_replica="0"} 790253500
# HELP monovtx_pending A Gauge to keep track of the total number of pending messages for the monovtx.
# TYPE monovtx_pending gauge
monovtx_pending{period="15m"} 264
monovtx_pending{period="1m"} 300
monovtx_pending{period="5m"} 285
# EOF
```

File: pkg/mvtxdaemon/server/service/rater/rater_test.go [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+func (m *raterMockHttpClient) Get(url string) (*http.Response, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if url == "https://p-mv-0.p-mv-headless.default.svc:2469/metrics" {
+		m.podOneCount = m.podOneCount + 20
+		resp := &http.Response{
+			StatusCode: 200,
+			// the test uses an abstract vertex without specifying vertex type, meaning it's neither source nor reduce,
+			// hence the default forwarder metric name "forwarder_data_read" is used to retrieve the metric
```
Comment (KeranYang) [inline pkg/mvtxdaemon/server/service/rater/rater_test.go:49]: please update the comment.

File: pkg/metrics/metrics.go [hunk 13]
```diff
@@ -28,10 +28,10 @@ const (
 	LabelISBService         = "isbsvc"
 	LabelPipeline           = "pipeline"
 	LabelVertex             = "vertex"
-	LabelMonoVertex         = "mvtx"
 	LabelVertexReplicaIndex = "replica"
 	LabelVertexType         = "vertex_type"
 	LabelPartitionName      = "partition_name"
+	LabelMonoVertex         = "mvtx"
```
Comment (KeranYang) [inline pkg/metrics/metrics.go:34]:
```suggestion
	LabelMonoVertexName         = "mvtx_name"
```

I saw you tested the entire PR using a real mono vertex, I am curious why the test looks good with the label name not matching with rust code.

File: rust/monovertex/src/metrics.rs [hunk 14]
```diff
@@ -17,22 +16,33 @@ use tokio::task::JoinHandle;
...
 use crate::transformer::TransformerClient;
 
 // Define the labels for the metrics
-pub const MONO_VERTEX_NAME: &str = "vertex";
-pub const REPLICA_LABEL: &str = "replica";
-pub const PARTITION_LABEL: &str = "partition_name";
-pub const VERTEX_TYPE_LABEL: &str = "vertex_type";
+pub const MONO_VERTEX_NAME_LABEL: &str = "mvtx_name";
+pub const REPLICA_LABEL: &str = "mvtx_replica";
+const PENDING_PERIOD_LABEL: &str = "period";
```
[the remark this answers is not in the document]
Comment (KeranYang) [reply rust/monovertex/src/metrics.rs:34]: same for the labels below.

File: pkg/mvtxdaemon/server/service/rater/rater.go [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+	"github.com/numaproj/numaflow/pkg/apis/numaflow/v1alpha1"
+	"github.com/numaproj/numaflow/pkg/shared/logging"
+	sharedqueue "github.com/numaproj/numaflow/pkg/shared/queue"
+)
+
+const CountWindow = time.Second * 10
+const MonoVtxReadMetricName = "monovtx_read_total"
+
+// MonoVtxRatable is the interface for the Rater struct.
+type MonoVtxRatable interface {
```
Comment (Krithika3) [inline pkg/mvtxdaemon/server/service/rater/rater.go:39]: Just a question out of curiosity, how is this vertex rater different from the pipeline rater? Thanks

File: pkg/metrics/metrics.go (hunk 13 above)
Comment (kohlisid) [reply pkg/metrics/metrics.go:34]: In our case there is no filtering on labels. Hence, we are able to retrieve the metric vals

File: pkg/mvtxdaemon/server/service/rater/rater.go (hunk 15 above)
Comment (kohlisid) [reply pkg/mvtxdaemon/server/service/rater/rater.go:39]: In the pipeline rater we calculate the rate per partition per pod.
In this scenario, we do not have such a concept, hence each pod just has a single metric value.

**Review by KeranYang — APPROVED** [review]
LGTM! Thanks @kohlisid !

**Review by whynowy — APPROVED** [review]

**Comment by vigith** [comment]
plesae give me till 05:00pm PDT today to review the Rust code.

**Pull request merged**